src/test/drop_select_packets.rs

```diff
@@ -0,0 +1,157 @@
+use bytes::BytesMut;
+use env_logger;
+use future_utils::FutureExt;
+use futures::future;
+use futures::{Future, Stream};
+use netsim;
+use netsim::wire::{IntoIpPlug, IpPacket, IpPlug, Ipv4Payload};
+use netsim::{Ipv4AddrExt, Ipv4Range, Ipv4Route};
+use packet::Packet;
+use rand;
+use std;
+use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+use tokio_core::reactor::Core;
+use tokio_io;
+use void::ResultVoidExt;
+use UtpSocket;
+
+#[test]
+fn test() {
+    for i in 0..6 {
+        drop_select_packets(i, true);
+        drop_select_packets(i, false);
+    }
+}
+
+fn drop_select_packets(n: u32, sender: bool) {
+    trace!("");
+    trace!("");
+    trace!("drop_select_packets({}, {})", n, sender);
+    let _ = env_logger::init();
+
+    let mut core = unwrap!(Core::new());
+    let handle = core.handle();
+    let network = netsim::Network::new(&handle);
+    let network_handle = network.handle();
+
+    let (addr_tx, addr_rx) = std::sync::mpsc::channel();
+
+    let res = core.run(future::lazy(move || {
+        let (mut plug_a, mut plug_b) = IpPlug::new_pair();
+
+        let mut i = 0;
+        let filter = move |packet| {
+            let prev_i = i;
+            i += 1;
+            if prev_i == n {
+                if let IpPacket::V4(ref ipv4_packet) = packet {
```

Comment:
That's cool.
How do you feel about implementing some helpers for `IpPlug` though? Say `plug.filter_ipv4().filter_udp()` could return stream that yields `Ipv4Payload::Udp`.
Then we could have an extension that implements `filter_utp()` inside this crate...

Reply:
Hmm, what would the API look like exactly?